### PR TITLE
247 allow filtering of chart participants

### LINKED ISSUE #247: Allow filtering of chart participants
### User stories

As a user on the Chart Page
When I filter by CHR or HC
Then only my filtered results are shown

As a user on the Chart Page
When I filter by included or excluded
Then only my filtered results are shown

### Implementation notes

Filtering will work by identifying participants from the `form_inclusionexclusion_criteria_review` assessment who meet the filter criteria and then only using those participants in the chart data.

**Filter assessment:**

`form_inclusionexclusion_criteria_review`

**Filter variables:**

`chrcrit_part` -> values can be `1` (CHR), `2` (HC), `3` (missing), both, or none

`included_excluded` -> values can be `1` (included), `2` (excluded), `3` (missing), both, or none

**Handling the filter data in the API**

When the API receives a request to filter chart data, we need to do the following:

1) Identify the participants matching the filter criteria
2) Update the [`allSubjects`](https://github.com/AMP-SCZ/dpdash/blob/pnl-devel/server/controllers/chartsController/index.js#L23-L32) logic to scope the subjects to only those matching the filter criteria

### Todo:

* Add a `ChartFilterForm` on the `ChartPage` that displays checkboxes for the filter values and a "Filter" submit button. The label for `chrcrit_part` is "Cohort", the label for `included_excluded` is "Incl/Excl".
* Expose an API endpoint that receives a GET request for chart data and returns the data needed to render a chart (the same way the current endpoint does).
* Send the filter form data as query params to the new API endpoint
* Handle the filter form query params in the API
* By default, all criteria should be selected

### PR DESCRIPTION
**Pull request by Elkrival**
[Video of Filters Feature](https://drive.google.com/file/d/1r9aeQA7O27U6kjv1SjKabiuVNj0fnGw_/view?usp=sharing)

Server:
* The server parses a querystring using the qs library then the results are passed to the toggleChartFilters function, this is to convert the string values into boolean values
* The filters go through a deep equality check against our default chart filters object, this is to apply the filters in the query or stay with the default functionality.
* The mongoQueryFromFilters function returns an array of objects, this is because if we query with an empty array without using the mongodb $or operator it will yield zero results even if a document matches other criteria
* The results return a cursor that we iterate and verify if the subject matches any of our filter criteria
* The results are applied and we perform the same calculations to render charts

Frontend:
* Added ChartFilterForm Component
* Updated routes and apply query params using qs library
* Added new styles and themes
* To turn the Checkboxes on and off we have to use a boolean, when the boolean is pased to qs it is converted into a string, in the backend we convert the string value into a boolean
* Used a drawer component to hide and show filters
* Added object to give a name and categorize filters

Screen Shot 2023-01-10 at 3 14 03 PM

### Message when no data is available for a chart

Screen Shot 2023-01-10 at 2 22 54 PM

### CODE REVIEW

File: constants/styles.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ export const colors = {
   deep_saffron: '#F89235',
   transparentWhite: 'rgba(255, 255, 255, 0.9)',
   columbia_blue: '#c0d9e1',
+  mgh_blue: '#089ca4',
```
Comment (mikestone14) [inline constants/styles.js]: do we need to add a new hex color for this feature?

File: server/controllers/chartsController/helpers.js [hunk 2]
```diff
@@ -246,3 +243,44 @@ export const postProcessData = (data, studyTotals) => {
 
   return processedDataBySite
 }
+
+export const mongoQueryFromFilters = (filters) => {
+  if (!filters) {
+    return
+  }
+
+  const chrCritFilters = filters.chrcrit_part.filter((f) => f.value === true)
```
Comment (mikestone14) [inline server/controllers/chartsController/helpers.js]:
```suggestion
  const chrCritFilters = filters.chrcrit_part.filter((f) => f.value)
```

File: server/controllers/chartsController/helpers.js [hunk 3]
```diff
@@ -246,3 +243,44 @@ export const postProcessData = (data, studyTotals) => {
...
+  const includedExcludedFilters = filters.included_excluded.filter(
+    (f) => f.value === true
```
Comment (mikestone14) [inline server/controllers/chartsController/helpers.js]:
```suggestion
    (f) => f.value
```

File: server/controllers/chartsController/helpers.js [hunk 4]
```diff
@@ -246,3 +243,44 @@ export const postProcessData = (data, studyTotals) => {
...
+  return Object.keys(filters).reduce((sanitizedFilters, filter) => {
+    sanitizedFilters[filter] = filters[filter].map((filter) => {
+      return {
+        ...filter,
+        value: filter.value === 'true' ? true : false,
+      }
+    })
+
+    return sanitizedFilters
+  }, {})
```
Comment (mikestone14) [inline server/controllers/chartsController/helpers.js]: would this function be easier/not needed if we used 0 and 1 for the filter values instead of true and false? I'm wondering why we need this vs being able to use the query params directly.

File: server/controllers/chartsController/helpers.test.js [hunk 5]
```diff
@@ -601,11 +602,69 @@ describe('chartsController - helpers', () => {
...
+        included_excluded: [
+          { name: 'Included', value: false },
+          { name: 'Excluded', value: true },
+          { name: 'Missing', value: false },
+        ],
+      })
+    })
+  })
+
+  describe(helpers.mongoQueryFromFilters, () => {
```
Comment (mikestone14) [inline server/controllers/chartsController/helpers.test.js:613]: great tests 🚀

File: server/controllers/chartsController/index.js [hunk 6]
```diff
@@ -1,80 +1,156 @@
-import { ObjectID } from 'mongodb'
-
+import { ObjectId } from 'mongodb'
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js:1]: I know we've had issues with the casing of this import and the version of the mongodb package in the past. Will this break on staging/production?

File: server/controllers/chartsController/index.js [hunk 7]
```diff
@@ -1,80 +1,156 @@
...
+export const graphDataController = async (
+  dataDb,
+  userAccess,
+  chart_id,
+  parsedQueryParams
+) => {
   const labelMap = new Map()
   const dataMap = new Map()
+  const areFiltersRequested =
+    !!Object.keys(parsedQueryParams).length &&
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: don't we only want to look at the `filters` query params? Let's leave the option available to have query params unrelated to filters.

File: server/controllers/chartsController/index.js [hunk 8]
```diff
@@ -1,80 +1,156 @@
...
+    !deepEqual(DEFAULT_CHART_FILTERS, sanitizeQueryParams(parsedQueryParams))
+  const filters = areFiltersRequested
+    ? sanitizeQueryParams(parsedQueryParams)
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: we're calling `sanitizeQueryParams` twice, could be DRY this up? Also, it'd be great to not have to sanitize the parsed query params. I think it's worth changing the value of the filter query param if it means we don't need this function.

File: server/controllers/chartsController/index.js [hunk 9]
```diff
@@ -1,80 +1,156 @@
...
-      { projection: { collection: 1, study: 1, _id: 0 } }
-    )
-    .toArray()
-
+    .findOne({ _id: ObjectId(chart_id) })
+  const allSubjects = await getAllSubjects({
+    dataDb,
+    chart,
+    userAccess,
+    areFiltersRequested,
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: do we need to pass this in to the function? What if `filters` was set to `undefined` if there are no filters or if all filters are selected? Then we can determine if we need to filter the subjects based on that value of `filters`.

File: server/controllers/chartsController/index.js [hunk 10]
```diff
@@ -1,80 +1,156 @@
...
   )
   const studyTotals = generateStudyTargetTotals(chart, allowedStudies)
 
-  for await (const subject of allSubjects) {
-    const { study } = subject
-    const subjectDayData = await dataDb
-      .collection(subject.collection)
-      .find({})
-      .toArray()
+  if (allSubjects.length) {
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: do we need this check? if `allSubjects` is an empty array that seems like it would be fine to use as we had this before.

File: server/controllers/chartsController/index.js [hunk 11]
```diff
@@ -1,80 +1,156 @@
...
+    }
+
+    return await dataDb
+      .collection(collections.toc)
+      .find(
+        {
+          assessment: chart.assessment,
+          subject: { $in: filteredSubjects },
+        },
+        { projection: { collection: 1, study: 1, _id: 0 } }
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: can we DRY up some of these options? It seems like we could extract constants for default options being the assessment and projection.

File: server/controllers/chartsController/index.js [hunk 12]
```diff
@@ -1,80 +1,156 @@
...
+  } else {
+    const mongoFilterQuery = mongoQueryFromFilters(filters)
+    const filteredSubjects = []
+    const criteriaCursor = await dataDb.collection(collections.toc).find(
+      {
+        assessment: INCLUSION_EXCLUSION_CRITERIA_FORM,
+        study: { $in: userAccess, $nin: STUDIES_TO_OMIT },
+      },
+      {
+        projection: { collection: 1, study: 1, _id: 0, subject: 1 },
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: do we need the collection and study or just the subject?

File: server/controllers/chartsController/index.js [hunk 13]
```diff
@@ -1,80 +1,156 @@
...
+        .findOne({ $or: mongoFilterQuery })
+
+      if (data) filteredSubjects.push(data?.subjectid)
+    }
+
+    return await dataDb
+      .collection(collections.toc)
+      .find(
+        {
+          assessment: chart.assessment,
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: should we restrict the study here too like the other queries?

File: server/controllers/chartsController/index.js [hunk 14]
```diff
@@ -1,80 +1,156 @@
...
+        projection: { collection: 1, study: 1, _id: 0, subject: 1 },
+      }
+    )
+
+    for await (let doc of criteriaCursor) {
+      const data = await dataDb
+        .collection(doc.collection)
+        .findOne({ $or: mongoFilterQuery })
+
+      if (data) filteredSubjects.push(data?.subjectid)
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]:
```suggestion
      if (data) filteredSubjects.push(data.subjectid)
```

File: server/controllers/chartsController/index.js [hunk 15]
```diff
@@ -1,80 +1,156 @@
...
+    }
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: if we call `toArray()` on the `criteriaCursor` request could we make it so we're not pushing onto the `filteredSubjects` array? Is it a bad idea to call `toArray` for performance reasons or otherwise? If we did change it to an array I think we would be able to change this to something like:

```ts
const filteredSubjects = await Promise.all(criteriaCursor.map(doc => {
  const data = await dataDb
    .collection(doc.collection)
    .findOne({ $or: mongoFilterQuery })

  return data?.subjectid
}))
```

Though we would then need to `filteredSubjects.filter(Boolean)` to get rid of empty items. 🤷

File: server/routes/charts.js [hunk 16]
```diff
@@ -46,12 +47,14 @@ router.route('/charts/:chart_id').get(ensureAuthenticated, async (req, res) => {
     const { dataDb } = req.app.locals
     const { chart_id } = req.params
     const { userAccess } = req.session
+    const chartFilters = qs.parse(req.query)
```
Comment (mikestone14) [inline server/routes/charts.js]: these are not chart filters but parsed query params.

```suggestion
    const parsedQueryParams = qs.parse(req.query)
```

File: views/ViewChart.react.js [hunk 17]
```diff
@@ -1,24 +1,49 @@
...
+import ChartFilterForm from './forms/CharFilterForm'
 
 const ViewChart = ({ graph, classes }) => {
-  const { title, description } = graph
+  const { title, description, filters } = graph
+  const [showFilterMenu, setFilterMenuToggle] = useState(false)
+  const openFilterMenu = () => setFilterMenuToggle(true)
+  const onCloseFilterMenu = () => setFilterMenuToggle(false)
+  const handleSubmit = (updatedFilters) =>
+    window.location.assign(routes.chart(graph.chart_id, updatedFilters))
```
Comment (mikestone14) [inline views/ViewChart.react.js]: we should nest the query params for filters in a `filters` object.

File: views/forms/CharFilterForm/index.jsx [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+import { FILTER_CATEGORIES } from '../../../constants'
+
+const ChartFilterForm = ({ initialValues, onSubmit, classes }) => {
+  const [values, setValues] = useState(initialValues)
+  const onChange = (target, filterKey) => {
+    setValues({
+      ...values,
+      [filterKey]: values[filterKey].map((filter) =>
+        filter.name === target.name
+          ? { ...filter, value: target.checked }
```
Comment (mikestone14) [inline views/forms/CharFilterForm/index.jsx]: should we be using `target.value` instead of `target.checked`?

File: views/forms/CharFilterForm/index.jsx [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+                      inset
+                    />
+                    <Checkbox
+                      checked={filter.value}
+                      onChange={({ target }) => onChange(target, filterKey)}
+                      name={filter.name}
+                      id={filterID}
+                      value={filter.value}
+                    />
+                  </ListItem>
```
Comment (mikestone14) [inline views/forms/CharFilterForm/index.jsx]: should these list items render labels with checkboxes inside? I'm thinking the resulting HTML should look something like:

```html
<ul>
  <li>
    <label>
      HC
      <input type="checkbox" checked="true" value="1" name="hc" id="inclusion-hc" />
    </label>
  </li>
</ul>

File: views/forms/CharFilterForm/index.jsx [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+      [filterKey]: values[filterKey].map((filter) =>
+        filter.name === target.name
+          ? { ...filter, value: target.checked }
+          : filter
+      ),
+    })
+  }
+
+  return (
+    <>
```
Comment (mikestone14) [inline views/forms/CharFilterForm/index.jsx]: I think technically this should be a form component.

File: views/forms/CharFilterForm/index.jsx [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+                    />
+                  </ListItem>
+                </List>
+              )
+            })}
+          </List>
+        )
+      })}
+      <Button
+        onClick={() => onSubmit(values)}
```
Comment (mikestone14) [inline views/forms/CharFilterForm/index.jsx]: if we make this component a form component we should probably make this a submit button. In that case we would either want `onSubmit(values)` to be called as part of an `onSubmit` prop on the form component or call `event.preventDefault()` on this `onClick` prop.

**Review by mikestone14 — COMMENTED** [review]
This looks good, left you some comments. In terms of the UI, did the MGB team ask you to put the form in a sidebar? I thought this was going onto the page itself.

File: server/controllers/chartsController/helpers.js (hunk 4 above)
Comment (Elkrival) [reply server/controllers/chartsController/helpers.js]: The Checkbox component uses the checked property to set the on or off state. The checked property only recognizes true or false as the appropriate values, additionally the qs library converts the boolean into a string.

File: server/controllers/chartsController/index.js (hunk 10 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: For some reason an empty array yields a blank page and I thought that would confuse users, by adding this it gives a message that there is no data available for the chart and it could prompt the users to select different filters, etc..

File: server/controllers/chartsController/index.js (hunk 12 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: With the current data structure we need the collection property 😦

File: server/controllers/chartsController/helpers.test.js (hunk 5 above)
Comment (Elkrival) [reply server/controllers/chartsController/helpers.test.js:613]: Thank you 😄

File: server/controllers/chartsController/helpers.js (hunk 4 above)
Comment (Elkrival) [reply server/controllers/chartsController/helpers.js]: This is what the query params look like server side
`{
  chrcrit_part: [
    { name: 'HC', value: 'true' },
    { name: 'CHR', value: 'false' },
    { name: 'Missing', value: 'true' }
  ],
  included_excluded: [
    { name: 'Included', value: 'true' },
    { name: 'Excluded', value: 'false' },
    { name: 'Missing', value: 'true' }
  ]
}`

Comment (mikestone14) [reply server/controllers/chartsController/helpers.js]: Right, I believe the `Checkbox` component behaves similarly to an `input` of type `checkbox`. We don't need to use the `checked` attribute in setting the filter, though. We can use the `value` and we can set the `value` to whatever we want.

File: constants/styles.js (hunk 1 above)
Comment (Elkrival) [reply constants/styles.js]: I think so, the other hex colors are used in the charts, but the app doesn't have a theme or color scheme. This is a color for the button that submits the filters. Would you like to have it removed?

File: server/controllers/chartsController/index.js (hunk 6 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js:1]: It shouldn't break anything in production

File: server/controllers/chartsController/index.js (hunk 9 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: We need it or something similar so could use your thoughts for this scenario.
When all the filters are selected we want the default behavior, but if there are subjects that have not taken the form_inclusionexclusion_criteria_review assessment, then we would be missing those folks from the counts.

The boolean here, has a deep equality check that if all filters are selected then we want the default behavior path.

What is your perspective of this scenario?

Comment (mikestone14) [reply server/controllers/chartsController/index.js]: I think it will be simpler to just pass the filters parameter and then use the default behavior if `filters` is `undefined` or if `filters` equals `DEFAULT_CHART_FILTERS`.

File: server/controllers/chartsController/index.js (hunk 10 above)
Comment (mikestone14) [reply server/controllers/chartsController/index.js]: can we look into why we get a blank page when allSubjects is empty? That is unexpected to me.

File: server/controllers/chartsController/index.js (hunk 15 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: There are a few reasons why I decided to go against using toArray.
1. toArray() needs to complete all of the lookups and then converts it into an array 
2. it is less performant 
3. We would be converting all of the results so we would still need to iterate the results and filter the ones that don't have data

In your example is that the Mongodb map? or the Array.map?

File: server/controllers/chartsController/index.js [hunk 22]
```diff
@@ -1,80 +1,156 @@
...
+      chart,
+      dataBySite: Array.from(dataBySite.values()),
+      labels: Array.from(labelMap.values()),
+      studyTotals,
+      filters,
+    }
+  } else {
+    return {
+      chart,
+      dataBySite: [],
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]: `Array.from(dataBySite.values())` when `dataBySite` is it's initial value (`new Map()`) is an empty array. Do we need to hard-code this or can we keep the return value here the same as the return value above on line 88?

File: views/forms/CharFilterForm/index.jsx (hunk 18 above)
Comment (Elkrival) [reply views/forms/CharFilterForm/index.jsx]: If we use a target.value we need a helper to convert the string value to a boolean

File: server/controllers/chartsController/index.js (hunk 15 above)
Comment (mikestone14) [reply server/controllers/chartsController/index.js]: ok I think your explanation makes sense and we should continue to do it this way. Thanks for the additional info!

File: server/controllers/chartsController/index.js (hunk 13 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: `    const criteriaCursor = await dataDb.collection(collections.toc).find(
      {
        assessment: INCLUSION_EXCLUSION_CRITERIA_FORM,
        study: { $in: userAccess, $nin: STUDIES_TO_OMIT },
      },
      {
        projection: ALL_SUBJECTS_MONGO_PROJECTION,
      }
    )`

Fortunately for us the prior cursor takes care of that for us, the results are subjects from sites/studies that a user already has access to. Instead we ask for the results of the assessment matching those subjects.

`subject: { $in: filteredSubjects },`

File: server/controllers/chartsController/index.js (hunk 22 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: views/forms/CharFilterForm/index.jsx (hunk 21 above)
Comment (Elkrival) [reply views/forms/CharFilterForm/index.jsx]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: views/forms/CharFilterForm/index.jsx (hunk 20 above)
Comment (Elkrival) [reply views/forms/CharFilterForm/index.jsx]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: views/forms/CharFilterForm/index.jsx (hunk 19 above)
Comment (Elkrival) [reply views/forms/CharFilterForm/index.jsx]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: views/forms/CharFilterForm/index.jsx (hunk 18 above)
Comment (Elkrival) [reply views/forms/CharFilterForm/index.jsx]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: views/ViewChart.react.js (hunk 17 above)
Comment (Elkrival) [reply views/ViewChart.react.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/routes/charts.js (hunk 16 above)
Comment (Elkrival) [reply server/routes/charts.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/controllers/chartsController/index.js (hunk 14 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/controllers/chartsController/index.js (hunk 11 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/controllers/chartsController/index.js (hunk 10 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/controllers/chartsController/index.js (hunk 8 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/controllers/chartsController/index.js (hunk 7 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/controllers/chartsController/helpers.js (hunk 4 above)
Comment (Elkrival) [reply server/controllers/chartsController/helpers.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/controllers/chartsController/helpers.js (hunk 3 above)
Comment (Elkrival) [reply server/controllers/chartsController/helpers.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: server/controllers/chartsController/helpers.js (hunk 2 above)
Comment (Elkrival) [reply server/controllers/chartsController/helpers.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/147bf8a9965b532f34bb176e1e651f339c9bf387

File: views/ViewChart.react.js [hunk 23]
```diff
@@ -1,26 +1,54 @@
...
+      <Drawer anchor="left" open={showFilterMenu} onClose={onCloseFilterMenu}>
+        <ChartFilterForm
+          initialValues={filters}
+          onSubmit={handleSubmit}
+          classes={classes}
+        />
+      </Drawer>
       <BarGraph graph={graph} classes={classes} />
-      <GraphTable graph={graph} />
+      {!!graph.dataBySite.length ?? <GraphTable graph={graph} />}
```
Comment (mikestone14) [inline views/ViewChart.react.js]: why is this needed? I don't think ?? is what we want here since the left side will never be null or undefined (it will be a boolean). So in this case I believe the graph table is never rendered?

File: views/forms/CharFilterForm/index.jsx [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+              ...filter,
+              value: filter.value === TRUE_STRING ? FALSE_STRING : TRUE_STRING,
+            }
+          : filter
+      ),
+    })
+  }
+
+  return (
+    <Form handleSubmit={(e) => onSubmit(e, values)}>
```
Comment (mikestone14) [inline views/forms/CharFilterForm/index.jsx]: do we need to submit the form event or can we just submit the values?

```suggestion
    <Form handleSubmit={() => onSubmit(values)}>
```

Comment (Elkrival) [reply views/forms/CharFilterForm/index.jsx]: ~~We need to submit the form event, but I can move it a level up so it's more like.~~

~~`(e) => (e.preventDefault(), onSubmit(formValues)`~~

The program did not like that, have to keep the e.preventDefault() before routing the page

File: server/controllers/chartsController/index.js (hunk 6 above)
Comment (mikestone14) [reply server/controllers/chartsController/index.js:1]: so `mongodb` exports both `ObjectID` and `ObjectId` and they are the same thing?

File: views/ViewChart.react.js (hunk 23 above)
Comment (Elkrival) [reply views/ViewChart.react.js]: Thank you. This is what it's supposed to be. https://github.com/AMP-SCZ/dpdash/pull/248/commits/1a688fa3ed46d74c7d2545e89cc2c27c0130f3fe

If we don't use this then the page looks like this 
Screen Shot 2023-01-11 at 4 05 40 PM

File: server/controllers/chartsController/index.js (hunk 6 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js:1]: yes they do the same thing. ObjectID is marked as deprecated and they want us to use ObjectId

Screen Shot 2023-01-11 at 4 02 17 PM

File: constants/styles.js (hunk 1 above)
Comment (Elkrival) [reply constants/styles.js]: https://github.com/AMP-SCZ/dpdash/pull/248/commits/2467f5a77a481233a7e141d858b7cd0668600c38

File: views/ViewChart.react.js (hunk 23 above)
Comment (Elkrival) [reply views/ViewChart.react.js]: Now it looks like this.
Screen Shot 2023-01-11 at 5 24 09 PM

File: server/controllers/chartsController/index.js (hunk 6 above)
Comment (mikestone14) [reply server/controllers/chartsController/index.js:1]: oh great, thanks for the explanation!

File: views/forms/CharFilterForm/index.jsx (hunk 24 above)
Comment (mikestone14) [reply views/forms/CharFilterForm/index.jsx]: I think we could still do it as

```tsx
<Form
  handleSubmit={(e) => {
    e.preventDefault()

    return onSubmit(values)
  }}
>
```

This way the parent component doesn't need to know about the event and can just handle the form values.

Comment (Elkrival) [reply views/forms/CharFilterForm/index.jsx]: Thank you for this
https://github.com/AMP-SCZ/dpdash/pull/248/commits/145b0275bc16b6a640f0184569937307c95b5ac5

**Review by mikestone14 — APPROVED** [review]
